### PR TITLE
DO NOT MERGE: RHOARDOC-1604: Reactive resolution

### PR DESCRIPTION
**Pull request by tradej**
Contains commits from #1300 and #1306. 

Both #1300 and #1306 should be closed without merging.

### CODE REVIEW

**Comment by tradej** [comment]
@huntley, please provide a peer review.

File: docs/topics/assembly_developing-and-deploying-spring-reactive-application.adoc [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+include::proc_using-oauth2-with-reactive-http-web-service-webflux.adoc[leveloffset=+1]
+
+include::proc_creating-reactive-smtp-email-app.adoc[leveloffset=+1]
+
+//concept module added
+include::con_server-sent-events.adoc[leveloffset=+1]
+include::proc_using-server-sent-events-spring-reactive-app.adoc[leveloffset=+1]
+
+//In addition to xref:example-rest-http-spring-boot[using an example], you can create new {runtime} applications from scratch and deploy them to OpenShift.
```
Comment (tradej) [inline docs/topics/assembly_developing-and-deploying-spring-reactive-application.adoc]: Please remove comments.

Unaddressed TODOs?

File: docs/topics/con_server-sent-events.adoc [hunk 2]
```diff
@@ -0,0 +1,9 @@
+= Server-Sent Events
```
Comment (tradej) [inline docs/topics/con_server-sent-events.adoc]: Anchor ID missing

File: docs/topics/con_spring-reactive-introduction.adoc [hunk 3]
```diff
@@ -0,0 +1,37 @@
+[id="spring-reactive-introduction_{context}"]
```
Comment (tradej) [inline docs/topics/con_spring-reactive-introduction.adoc]: ID doesn't match the heading

File: docs/topics/con_spring-reactive-introduction.adoc [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+In addition to all the annotation types inherited from Spring MVC, WebFlux also supports `@RequestBody` arguments.
+
+//Efficient resource usage::
+//Reactor and WebFlux handle high volumes of concurrent events (on the order of tens of millions of messages per second) using a smaller amount of threads compared to Servlet-based libraries handling the same workload.
+
+Functional programming support::
+Reactor interacts directly with the Java 8 Functional API, as well as `CompletablebFuture`, and `Stream` APIs.
+In addition to annotation-based endpoints, WebFlux also supports functional endpoints.
+
+//TODO: add additional resources section
```
Comment (tradej) [inline docs/topics/con_spring-reactive-introduction.adoc]: Unaddressed TODOs?

File: docs/topics/con_spring-reactive-introduction.adoc [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+Additional advantages of using Spring WebFlux include:
+
+Similarity with SpringMVC::
+Due to the high degree of similarity between the SpringMVC API types and WebFlux API types, it is easy to re-apply knowledge of SpringMVC to programming applications with WebFLux.
+
+Annotated controllers::
+WebFlux retains the endpoint controller annotations introduced by SpringMVC (Both SpringMVC and WebFlux support reactive RxJava2 and Reactor return types).
+In addition to all the annotation types inherited from Spring MVC, WebFlux also supports `@RequestBody` arguments.
+
+//Efficient resource usage::
```
Comment (tradej) [inline docs/topics/con_spring-reactive-introduction.adoc]: Why is this commented out?

File: docs/topics/con_spring-web-reactive.adoc [hunk 6]
```diff
@@ -0,0 +1,19 @@
+[id='spring-web-reactive_{context}']
```
Comment (tradej) [inline docs/topics/con_spring-web-reactive.adoc]: ID doesn't match the heading

File: docs/topics/proc_creating-a-reactive-http-service-with-webflux.adoc [hunk 7]
```diff
@@ -0,0 +1,83 @@
+[id='creating-a-reactive-spring-boot-http-service-with-webflux_{context}']
+= Creating a reactive {SpringBoot} HTTP service with WebFlux
+
+Create a basic reactive Hello World HTTP web service using {SpringBoot} and WebFlux.
+
+.Prerequisites
+
+* Java JDK 8 or Java JDK 11 installed
```
Comment (tradej) [inline docs/topics/proc_creating-a-reactive-http-service-with-webflux.adoc]: Don't we have a variable for it?

File: docs/topics/proc_creating-a-reactive-http-service-with-webflux.adoc [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+* Maven installed
+// this configuration might not apply entirely while examples still in tech preview
```
Comment (tradej) [inline docs/topics/proc_creating-a-reactive-http-service-with-webflux.adoc]: Mention this in the doc?

File: docs/topics/proc_creating-a-reactive-http-service-with-webflux.adoc [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+* A Maven-based application project configured link:{link-guide-spring-boot}#configuring-your-application-to-use-spring-boot_spring-boot[configured to use Spring Boot]
+
+
+
+.Procedure
+
+. Add `vertx-spring-boot-starter-http` as a dependency in the `pom.xml` file of your project.
+
+.`pom.xml`
```
Comment (tradej) [inline docs/topics/proc_creating-a-reactive-http-service-with-webflux.adoc]: This is not indented properly. Should be connected to L17 using `+`.

Same in multiple places.

File: docs/topics/proc_creating-a-reactive-http-service-with-webflux.adoc [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
++
+[source,bash,options="nowrap",subs="attributes+"]
+----
+$ curl http://localhost:8080/hello?name=John
+Hello, John!
+----
+
+.Additional resources
+
+You can link:{link-guide-spring-boot}#deploying-runtime-application-to-openshift_spring-boot[deploy your application to an OpenShift cluster] using Fabric8 Maven Plugin.
```
Comment (tradej) [inline docs/topics/proc_creating-a-reactive-http-service-with-webflux.adoc]: We use bullet points in Additional resources.

File: docs/topics/proc_creating-a-reactive-http-service-with-webflux.adoc [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+You can also configure your application for link:{link-guide-spring-boot}#deploying-runtime-application-to-standalone-red-hat-enterprise-linux_spring-boot[deployment on standalone Red{nbsp}Hat Enterprise Linux].
+
+
+// add links to: https://spring.io/guides/gs/reactive-rest-service/
```
Comment (tradej) [inline docs/topics/proc_creating-a-reactive-http-service-with-webflux.adoc]: Not finished?

File: docs/topics/proc_using-server-sent-events-spring-reactive-app.adoc [hunk 12]
```diff
@@ -0,0 +1,99 @@
+[id='using-server-sent-events-in-a-spring-reactive-application_{context}']
+= Using Server-Sent Events in a Spring Reactive application
+
+Create a simple service that request and receives a stream of server-sent events (SSE)
```
Comment (tradej) [inline docs/topics/proc_using-server-sent-events-spring-reactive-app.adoc]: Missing period.

File: docs/topics/con_server-sent-events.adoc [hunk 13]
```diff
@@ -0,0 +1,9 @@
...
+
+Server-sent events (SSE) is a protocol that you can use to push unidirectional events form an event source to a web-based or browser-based client over HTTP.
+It is more light-weight compared to REST or WebSockets, and it does not require the target client to be aware of the protocol being used.
```
[the remark this answers is not in the document]
Comment (Ladicek) [reply docs/topics/con_server-sent-events.adoc]: > It is more light-weight compared to REST or WebSockets, and it does not require the target client to be aware of the protocol being used.

This doesn't make much sense to me. Clearly the client needs to understand (be aware of) the SSE protocol. How much is it lighter-weight compared to REST -- I'm not sure, but the client needs to initiate an HTTP connection to an endpoint, which would typically be part of a REST architecture.

(I didn't go through the entire PR, but this snippet caught my attention in my GitHub notifications :-) )

File: docs/spring-boot-runtime/master.html [hunk 14]
```diff
@@ -0,0 +1,10708 @@
+<!DOCTYPE html>
```
[the remark this answers is not in the document]
Comment (inoxx03) [reply docs/spring-boot-runtime/master.html]: removed

File: docs/topics/assembly_developing-and-deploying-spring-reactive-application.adoc [hunk 15]
```diff
@@ -0,0 +1,25 @@
+[id='developing-reactive-spring-boot-applications-using-spring-boot-vertx_{context}']
```
[the remark this answers is not in the document]
Comment (inoxx03) [reply docs/topics/assembly_developing-and-deploying-spring-reactive-application.adoc]: Changed heading to match ID

File: docs/topics/assembly_developing-and-deploying-spring-reactive-application.adoc [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+= Developing reactive applications using {SpringReactive}
+
+:parent-context: {context}
+
```
[the remark this answers is not in the document]
Comment (inoxx03) [reply docs/topics/assembly_developing-and-deploying-spring-reactive-application.adoc]: introduction paragraph added

File: docs/topics/con_server-sent-events.adoc (hunk 13 above)
[the remark this answers is not in the document]
Comment (inoxx03) [reply docs/topics/con_server-sent-events.adoc]: @Ladicek thank you for pointing this out. I've re-worded the introduction and I will submit it with the rest of the required corrections in a separate PR which I will reference in the comments.

**Comment by inoxx03** [comment]
@tradej, @huntley and @Ladicek  thank you for the review comments, I am addressing them in PR #1309. This PR can be closed without merging.

File: docs/topics/assembly_developing-and-deploying-spring-reactive-application.adoc (hunk 1 above)
Comment (inoxx03) [reply docs/topics/assembly_developing-and-deploying-spring-reactive-application.adoc]: TODOs addressed and/or removed, respectively

File: docs/topics/con_server-sent-events.adoc (hunk 2 above)
Comment (inoxx03) [reply docs/topics/con_server-sent-events.adoc]: Anchor ID added

File: docs/topics/proc_creating-a-reactive-http-service-with-webflux.adoc (hunk 8 above)
Comment (inoxx03) [reply docs/topics/proc_creating-a-reactive-http-service-with-webflux.adoc]: According to Libor, the reactive webflux starters are going to be released under `snowdrop` and supported in 2.1.6. The configuration therefore applies. Comment removed.

File: docs/topics/proc_creating-a-reactive-http-service-with-webflux.adoc (hunk 11 above)
Comment (inoxx03) [reply docs/topics/proc_creating-a-reactive-http-service-with-webflux.adoc]: Added proper link, caption, and bullet point.

File: docs/topics/proc_creating-a-reactive-http-service-with-webflux.adoc (hunk 10 above)
Comment (inoxx03) [reply docs/topics/proc_creating-a-reactive-http-service-with-webflux.adoc]: Fixed. Additional resources links now formatted as bullet points.

File: docs/topics/con_server-sent-events.adoc [hunk 17]
```diff
@@ -0,0 +1,9 @@
+= Server-Sent Events
+
+Server-sent events (SSE) is a protocol that you can use to push unidirectional events form an event source to a web-based or browser-based client over HTTP.
```
[the remark this answers is not in the document]
Comment (inoxx03) [reply docs/topics/con_server-sent-events.adoc]: Thanks:) fixed.

File: docs/topics/proc_using-server-sent-events-spring-reactive-app.adoc (hunk 12 above)
Comment (inoxx03) [reply docs/topics/proc_using-server-sent-events-spring-reactive-app.adoc]: Thanks. fixed.

File: docs/topics/con_server-sent-events.adoc [hunk 18]
```diff
@@ -0,0 +1,9 @@
...
+It is more light-weight compared to REST or WebSockets, and it does not require the target client to be aware of the protocol being used.
+As opposed to WebsSockets, it does not handle bidirectional communication beetween an event source
```
[the remark this answers is not in the document]
Comment (inoxx03) [reply docs/topics/con_server-sent-events.adoc]: Thanks:)
added:

> ...and the client.

File: docs/topics/con_server-sent-events.adoc [hunk 19]
```diff
@@ -0,0 +1,9 @@
+= Server-Sent Events
+
+Server-sent events (SSE) is a protocol that you can use to push unidirectional events form an event source to a web-based or browser-based client over HTTP.
+It is more light-weight compared to REST or WebSockets, and it does not require the target client to be aware of the protocol being used.
+As opposed to WebsSockets, it does not handle bidirectional communication beetween an event source 
+The specification for SSE is incorporated into HTML5, and is widely supported by browsers, including legacy versions. It can can be easily used from the command line and is relatively simple to set up compared to other protocols.
```
[the remark this answers is not in the document]
Comment (inoxx03) [reply docs/topics/con_server-sent-events.adoc]: Changed _browsers_ to _web browsers_. (Good catch! Thanks.)
Rewrote the rest of the section, but the new formulation is consistent with your re-wording suggestions.

File: docs/topics/con_server-sent-events.adoc [hunk 20]
```diff
@@ -0,0 +1,9 @@
...
+SSE is suitable for use cases that require frequent updates from the server to the client, while updates from the client side to the server are expected to be less frequent and can therefore be handled over a different protocol, such as REST.
```
[the remark this answers is not in the document]
Comment (inoxx03) [reply docs/topics/con_server-sent-events.adoc]: @huntley Thank you for the suggestion. I divided up the sentences in a slightly different way, but I think the readability of the section improved nonetheless:

> SSE is suitable for use cases that require frequent updates from the server to the client, while updates from the client side to the server are expected to be less frequent.
Updates form the client side to the server can then be handled handled over a different protocol, such as REST.

WDYT?

File: docs/topics/con_server-sent-events.adoc [hunk 21]
```diff
@@ -0,0 +1,9 @@
...
+Examples of such use cases include live social media feed updates or notifications sent to a client when new files are uploaded to a file server.
```
[the remark this answers is not in the document]
Comment (inoxx03) [reply docs/topics/con_server-sent-events.adoc]: fixed, thanks!

File: docs/topics/con_spring-reactive-introduction.adoc [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+
+On top of the Reactor library, Spring Reactive relies on WebFlux, an asynchronous web application framework based on the Reactive Streams specifications.
+While WebFlux is designed to work primarily with Spring Reactor, due to its compliance with Reactive Streams, WebFlux can also operate in other asynchronous runtime environments, such as Netty or Undertow.
+
+Spring WebFlux and Reactor enable you to create applications that are:
+
+* Non-blocking, meaning that the application does not remain inaccessible to further requests while waiting for a response from a remote component or service that is required to complete handling the current request.
+* Asynchronous, meaning that rather than handling incoming requests in a sequence, the application responds to event from an event stream (also called an event loop) by generating its own response events and publishing them to the event stream where they can be picked up by other clients in the application and processed further.
```
[the remark this answers is not in the document]
Comment (inoxx03) [reply docs/topics/con_spring-reactive-introduction.adoc]: I like your wording better. Changed. Thank you.

File: docs/topics/proc_using-server-sent-events-spring-reactive-app.adoc [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+The service accepts a HTTP request and responds by returning a continuous stream of events until you cancel the request.
```
[the remark this answers is not in the document]
Comment (inoxx03) [reply docs/topics/proc_using-server-sent-events-spring-reactive-app.adoc]: Technically, it's just one request to open to establish the connection, the server then pushes the events to the client side over the same connection until that request is cancelled and the connection is closed.

I can reword as:

> The service accepts HTTP requests and responds by returning a continuous stream of events. The connection between the client and server remains open and the server continuously pushes new events to the client using the same connection.
Cancelling the request closes the connection and stops the stream.

WDYT?

File: docs/topics/proc_using-server-sent-events-spring-reactive-app.adoc [hunk 24]
```diff
@@ -0,0 +1,99 @@
...
+
+. Create the main class of your application:
+
+.`SseExampleApplication.java`
+[source,java,options="nowrap",subs="attributes+"]
+----
+include::resources/spring-boot/reactive-examples/example-sse-application.java[]
+----
+
+. Create a Server-sent Event controller class for your application.
```
[the remark this answers is not in the document]
Comment (inoxx03) [reply docs/topics/proc_using-server-sent-events-spring-reactive-app.adoc]: added `+` to preserve correct indentation of the code block

File: docs/topics/proc_using-server-sent-events-spring-reactive-app.adoc [hunk 25]
```diff
@@ -0,0 +1,99 @@
...
+  <dependencies>
+  ...
+include::resources/spring-boot/reactive-examples/example-sse-service-pom.xml[tags=sse-service-pom-dependencies]
+  ...
+  <dependencies>
+...
+</project>
+----
+
+. Create the main class of your application:
```
[the remark this answers is not in the document]
Comment (inoxx03) [reply docs/topics/proc_using-server-sent-events-spring-reactive-app.adoc]: added `+` to preserve correct indentation of the code block

File: docs/topics/proc_using-server-sent-events-spring-reactive-app.adoc (hunk 24 above)
[the remark this answers is not in the document]
Comment (inoxx03) [reply docs/topics/proc_using-server-sent-events-spring-reactive-app.adoc]: Fixed, including other instances of incorrect capitalization.

File: docs/topics/proc_using-server-sent-events-spring-reactive-app.adoc [hunk 26]
```diff
@@ -0,0 +1,99 @@
...
+[source,java,options="nowrap",subs="attributes+"]
+----
+include::resources/spring-boot/reactive-examples/example-sse-service-sse-controller.java[]
+----
+
+. OPTIONAL: Run and test your application locally 
+
+[source,bash,options="nowrap",subs="attributes+"]
+----
+$ cd myApp
```
[the remark this answers is not in the document]
Comment (inoxx03) [reply docs/topics/proc_using-server-sent-events-spring-reactive-app.adoc]: Good catch. Thank you. I included a proper instruction for this step.

File: docs/topics/proc_using-server-sent-events-spring-reactive-app.adoc [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+data:1338766210
+
+data:-666543077
+...
+----
++
+Press `Ctrl`+`C` to cancel your HTTP request and terminate the stream of responses.
+
+
+.Additional resources
```
[the remark this answers is not in the document]
Comment (inoxx03) [reply docs/topics/proc_using-server-sent-events-spring-reactive-app.adoc]: Added bullet points to entries in the _Addtional resources_ section.

**Comment by inoxx03** [comment]
This PR has been superseded by PR #1309. Closing without merging.

**Pull request closed without merging**